src/us/kbase/workspace/database/provenance/Contributor.java

```diff
@@ -0,0 +1,290 @@
+	 *
+	 * @param contributorType
+	 *                the type of contributor.
+	 * @param name    contributor name
+	 * @return the builder.
+	 */
+	public static Builder getBuilder(final ContributorType contributorType, final String name) {
+		final List<String> errorList = new ArrayList<>();
+		if (contributorType == null) {
+			errorList.add("contributorType cannot be null");
+		}
+		return new Builder(contributorType, name, errorList);
+	}
+
+	/**
+	 * Gets a builder for a {@link Contributor}.
+	 *
+	 * @param contributorType
+	 *                the type of contributor, as a string.
+	 * @param name    contributor name
+	 * @return the builder.
+	 */
+	public static Builder getBuilder(final String contributorType, final String name) {
+
+		final List<String> errorList = new ArrayList<>();
+		try {
+			return new Builder(ContributorType.getType(contributorType), name, errorList);
+		} catch (Exception e) {
+			errorList.add(e.getMessage());
+			return new Builder(null, name, errorList);
+		}
+	}
+
+	/** A builder for a {@link Contributor}. */
+	public static class Builder {
+
+		private ContributorType contributorType;
+		private String name;
+		private String creditName = null;
+		private String contributorID = null;
+		private List<Organization> affiliations = null;
+		private List<ContributorRole> contributorRoles = null;
+		private List<String> contributorRoleStrings = null;
+		private List<String> errorList;
+
+		private Builder(final ContributorType contributorType, final String name, final List<String> errorList) {
+			this.errorList = errorList;
+			this.contributorType = contributorType;
+			this.name = Common.processString(name);
+			if (this.name == null) {
+				errorList.add("name cannot be null or whitespace only");
+			}
+		}
+
+		public Builder withContributorID(final String contributorID) {
+			try {
+				this.contributorID = Common.checkPid(contributorID, "contributorID", true);
+			} catch (Exception e) {
+				this.errorList.add(e.getMessage());
+			}
+			return this;
+		}
+
+		public Builder withCreditName(final String creditName) {
+			this.creditName = Common.processString(creditName);
+			return this;
+		}
+
+		public Builder withAffiliations(final List<Organization> affiliations) {
+			this.affiliations = affiliations;
+			return this;
+		}
+
+		public Builder withContributorRoles(final List<ContributorRole> contributorRoles) {
+			this.contributorRoleStrings = null;
+			this.contributorRoles = contributorRoles;
+			return this;
+		}
+
+		public Builder withContributorRoleStrings(final List<String> contributorRoleStrings) {
+			this.contributorRoleStrings = contributorRoleStrings;
+			this.contributorRoles = null;
+			return this;
+		}
+
+		public Contributor build() {
+
+			if (creditName != null && contributorType == ContributorType.ORGANIZATION) {
+				errorList.add("the creditName field is only used with contributorType person");
+			}
+
+			if (contributorRoleStrings != null) {
```

Comment:
If there's more than one set of contrib role strings / contrib roles submitted, needless work is done.

<whinge>
I really dislike this constructor style that allows the same field to be set numerous times. It has no benefits for my use case (if a field was set more than once in the builder, it would suggest a programming error) and adds to the test burden. 
</whinge>

Reply:
> If there's more than one set of contrib role strings / contrib roles submitted, needless work is done.

But that's almost never going to happen. Better to make the code easier to understand and localize the processing to the input method, IMO

> I really dislike this constructor style that allows the same field to be set numerous times. It has no benefits for my use case (if a field was set more than once in the builder, it would suggest a programming error) and adds to the test burden.

Yeah, that is probably one of the most annoying things about the builder pattern. The benefits outweigh the drawbacks by a long way though:

* Adding optional fields easily vs winding up with more and more longer and longer constructors
* Readability when instantiating a new instance
* Providing alternate implementations of the class (e.g. for the PID class, the builder could return a class with only the pid field if the other 2 fields are empty, saving 8-16 bytes of data per instance)

It'd be nice if java could do something like python's keyword / default arguments.

Theoretically, the with* methods could throw an error if the instance variable is already set, but that also adds to the test burden and just seems really weird to me. It also means the calls are no longer idempotent

Reply:
> Theoretically, the with* methods could throw an error if the instance variable is already set, but that also adds to the test burden and just seems really weird to me. It also means the calls are no longer idempotent

I thought about doing that but it ends up being even more of a faff -- a "worst of both worlds" situation.

I would prefer to be able to use a hash/map/dict to initialise the object -- makes interconverting between objects a lot easier, there's no setting/resetting of fields, easy to add more fields as required, and generally much neater and sweeter. Grumble grumble...

Reply:
there's nothing stopping you from adding a static method to do that. I'm just not sure how much of a benefit it'd be given the incoming data will be in a class

Reply:
> > If there's more than one set of contrib role strings / contrib roles submitted, needless work is done.

> But that's almost never going to happen. Better to make the code easier to understand and localize the processing to the input method, IMO

I tried to guess which one you would be more likely to complain about: unnecessary work or having the processing in the `build` method. Obviously I picked wrong this time!

Reply:
CPU hours are a lot cheaper than engineering hours

Reply:
Moved the method

Reply:
Ugh, looking at it now it occurs to me that if the method is called once with a set of strings that causes errors the build will be permanently broken. If it was called again with an acceptable set of strings the build will be unable to complete, even though it's fine. My bad for not thinking this all the way through.

I can think of a couple fixes, up to you (or if you have a better fix)
* have a specific error list for this function and clear it when it's called
* revert back to the old implementation

Reply:
Where are these classes going to be used? Passing data back and forth between their spec-generated equivalent for interacting with the JSONAPI, and reading/writing to the DB? Anywhere else? It should all be in internal-to-the-workspace operations, correct? Is there any reason in either of those cases why one would _want_ to set the values for a field twice?

It could be argued that it would be helpful to accumulate these errors as they indicate (a) there's some dodgy input proposed for that field and (b) if the input that generated the error is unexpected, that there's a programming error somewhere.

Reply:
I mean if the builder is in a state such that the build is valid, then calling build() should work. The user definitely shouldn't get an error saying "input X is wrong" when input X is correct.

Reply:
I don't get why one would _want_ to set the values for a field twice, other than that it's an established design pattern and it doesn't matter if in this use case, we don't need that particular feature. Anyway, moved the code in question.

Reply:
It's rare, but I've done things like

```
builder.withThing(thing)
if (conditional)
    builder.withThing(otherthing)
```

... but I'd abandoned collecting errors way before I started using builders, so it doesn't cause the unique situation here

Reply:
:+1: